lib/puzzles/cubit/puzzles_cubit.dart

```diff
@@ -0,0 +1,163 @@
+import 'package:custom_platform_device_id/platform_device_id.dart';
+import 'package:equatable/equatable.dart';
+import 'package:hydrated_bloc/hydrated_bloc.dart';
+import 'package:json_annotation/json_annotation.dart';
+import 'package:puzzle_repository/puzzle_repository.dart';
+
+part 'puzzles_state.dart';
+part 'puzzles_cubit.g.dart';
+
+class PuzzlesCubit extends HydratedCubit<PuzzlesState> {
+  PuzzlesCubit({required PuzzleRepository puzzleRepository})
+      : _puzzleRepository = puzzleRepository,
+        super(const PuzzlesState());
+
+  final PuzzleRepository _puzzleRepository;
+
+  Future<void> verifyUser({required String username}) async {
+    try {
+      emit(
+        state.copyWith(
+          userVerificationStatus: UserVerificationStatus.verifying,
+        ),
+      );
+      final deviceId = await PlatformDeviceId.getDeviceId;
+      if (deviceId == null) {
+        emit(
+          state.copyWith(userVerificationStatus: UserVerificationStatus.failed),
+        );
+        return;
+      }
+      final user = await _puzzleRepository.verifyUser(
+        username: username,
+        deviceId: deviceId,
+      );
+      emit(
+        state.copyWith(
+          user: user,
+          userVerificationStatus: UserVerificationStatus.verified,
+        ),
+      );
+    } on UsernameAlreadyTaken catch (_) {
+      emit(
+        state.copyWith(
+          userVerificationStatus: UserVerificationStatus.existingUsername,
+        ),
+      );
+    } catch (_) {
+      emit(
+        state.copyWith(
+          userVerificationStatus: UserVerificationStatus.failed,
+        ),
+      );
+    }
+  }
+
+  Future<void> fetchPuzzles() async {
+    try {
+      emit(
+        state.copyWith(puzzlesFetchingStatus: FetchStatus.fetching),
+      );
+      final puzzles = await _puzzleRepository.fetchPuzzles(
+        userId: state.user!.userID,
+      );
+      emit(
+        state.copyWith(
+          puzzles: puzzles,
+          puzzlesFetchingStatus: FetchStatus.fetched,
+        ),
+      );
+    } catch (_) {
+      emit(state.copyWith(puzzlesFetchingStatus: FetchStatus.failed));
+    }
+  }
+
+  Future<void> submitAnswer({
+    required int puzzleId,
+    required dynamic answer,
+  }) async {
+    try {
+      emit(
+        state.copyWith(
+          currentPuzzleSubmissionStatus: SubmissionStatus.submitting,
+        ),
+      );
+      final userProgress = await _puzzleRepository.submitAnswer(
+        userId: state.user!.userID,
```

Comment:
Shouldn't we ensure user is not null instead of force unwrapping here?

Reply:
Hey @felangel, We're performing a check here to determine if the user is null. If the user is null, the UI for submitting quiz answers will not be displayed.

https://github.com/felangel/flutter_and_friends/blob/0cf7373868b4d383c8cbfad3e7d64263f03b4c8e/lib/puzzles/widgets/puzzles_page_content.dart#L46